### PR TITLE
[WIP] sys-kernel/vanilla-kernel: package to install built kernel

### PR DESCRIPTION
**Pull request by mgorny**
Here's an early proof-of-concept. It's still missing a lot and needs more work.

So far it unpacks the sources, runs `defconfig` (which should be replaced by something better), does `modules_prepare`, builds and installs the kernel. It also installs the sources + objects generated by `modules_prepare`.

TODO:
- [x] figure out how to install kernel properly
  - [x] build initramfs?
  - [x] and handle various bootloaders
    - [x] lilo
    - [x] grub
    - [x] systemd-boot
  - [x] <del>remount /boot</del> (mount-boot doesn't work anymore)
- [x] use savedconfig
- [ ] add custom-cflags?
- [x] figure out if we can minimize installed sources
  - [x] do we need all compiled objects, or just final tools?
  - [x] do we need all sources or just headers+makefiles?
- [x] handle switching to ld.bfd when other linker is selected
- [x] find a working .config

Bugs:
- [ ] `@module-rebuild` catches new kernels
- [x] `Module.symvers` is missing

### CODE REVIEW

**Comment by alexxy** [comment]
Hi!

I think defconfig can be replaced by user savedconfig. or something default. 
Q about objects and so on depend on wheather you want or not build extra modules. 

Some of such exampled were made by Calculate linux (they used gentoo as base)
Something like [this](https://old.calculate-linux.org/packages/sys-kernel/calculate-sources)

**Comment by mgorny** [comment]
Ok, I think it's ready for initial testing. It doesn't boot for me but it may be something wrong with my setup. It'd be nice if someone more experienced could look at it.

**Comment by Zlogene** [comment]
@mgorny please also add me to metadata

File: sys-kernel/vanilla-kernel/vanilla-kernel-5.4.1.ebuild [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+LICENSE="GPL-2"
+SLOT="${PV}"
+KEYWORDS="~amd64"
+IUSE="+initramfs"
+
+# install-DEPEND actually
+RDEPEND="sys-apps/debianutils
+	initramfs? ( sys-kernel/dracut )"
+# we steal default configs from genkernel
+BDEPEND="sys-kernel/genkernel"
```
Comment (Zlogene) [inline sys-kernel/vanilla-kernel/vanilla-kernel-5.4.1.ebuild]: do we really need to hard-depend on genkernel in order to steal their config? We can host it somewhere else (like devspace)

File: sys-kernel/vanilla-kernel/vanilla-kernel-5.4.1.ebuild [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+S=${WORKDIR}/${MY_P}
+
+LICENSE="GPL-2"
+SLOT="${PV}"
+KEYWORDS="~amd64"
+IUSE="+initramfs"
+
+# install-DEPEND actually
+RDEPEND="sys-apps/debianutils
+	initramfs? ( sys-kernel/dracut )"
```
Comment (Zlogene) [inline sys-kernel/vanilla-kernel/vanilla-kernel-5.4.1.ebuild]: genkernel can also generate initramfs, so we can provide it as an alternative here

Comment (mgorny) [reply sys-kernel/vanilla-kernel/vanilla-kernel-5.4.1.ebuild]: genkernel is horrible, so I'd rather not waste my time on figuring it out.

**Comment by Zlogene** [comment]
I have successfully booted my box with customc onfiguration put into savedconfig, the only thing I had to do is to copy compiled kernel into the dir where efidtub is searching for it and rename a kernel binary,

File: sys-kernel/vanilla-kernel/vanilla-kernel-5.4.1.ebuild (hunk 1 above)
Comment (mgorny) [reply sys-kernel/vanilla-kernel/vanilla-kernel-5.4.1.ebuild]: Resolved via fetching Arch's config.

**Comment by mgorny** [comment]
Ok, the newest version builds a working kernel+initramfs that boots with the default config. On the minus side, initramf is 50M in size but I suppose this is not something that can be avoided easily.

Now, does anyone else have some comments before I commit the initial version? That is, after getting debianutils/installkernel changes merged first, then updating it to the latest kernel version.

**Comment by Zlogene** [comment]
@mgorny What I was thinking about is dracut call. Default dracut installation is not optimal for good enough initrd. Should not we rather notify users with ewarn/eeror about the necessity of initrd being created if the default config was used?

**Comment by mgorny** [comment]
It defeats the goal of having it working out of the box with no extra effort. No other distro tells users to set up initramfs themselves after building the kernel.

File: profiles/base/package.use.force [hunk 3]
```diff
@@ -1,6 +1,13 @@
 # Copyright 1999-2019 Gentoo Authors
 # Distributed under the terms of the GNU General Public License v2
 
+# Michał Górny <mgorny@gentoo.org> (2019-12-06)
+# Non-initramfs builds do not work with the default .config.  Force
+# the mask, so that users don't brick their systems.  If somebody wants
+# to avoid initramfs, he needs to use a custom .config and unforce
+# the flag.
+sys-kernel/vanilla-kernel initramfs
```
Comment (SoapGentoo) [inline profiles/base/package.use.force]: But Dennie Ritchie didn't have an initramfs in 1979!

File: sys-kernel/installkernel-systemd-boot/installkernel-systemd-boot-0.ebuild [hunk 4]
```diff
@@ -0,0 +1,27 @@
+# Copyright 2019 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=7
+
+DESCRIPTION="Wrap kernel-install from systemd-boot as installkernel"
+HOMEPAGE="https://wiki.gentoo.org/wiki/No_homepage"
+SRC_URI=""
+S=${WORKDIR}
```
Comment (SoapGentoo) [inline sys-kernel/installkernel-systemd-boot/installkernel-systemd-boot-0.ebuild]: put this after `RDEPEND`

File: sys-kernel/vanilla-kernel/vanilla-kernel-5.4.1.ebuild [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+MY_P=linux-${PV}
+CONFIG_VER=5.4.1.arch1-1
+CONFIG_HASH=d6bbb261bd2c8f83b9c93041dae9aa4175e20e6e
+
+DESCRIPTION="Linux kernel built from vanilla upstream sources"
+HOMEPAGE="https://www.kernel.org/"
+SRC_URI="https://cdn.kernel.org/pub/linux/kernel/v5.x/${MY_P}.tar.xz
+	https://git.archlinux.org/svntogit/packages.git/plain/trunk/config?h=packages/linux&id=${CONFIG_HASH}
+		-> linux-${CONFIG_VER}.config"
+S=${WORKDIR}/${MY_P}
```
Comment (SoapGentoo) [inline sys-kernel/vanilla-kernel/vanilla-kernel-5.4.1.ebuild]: same

**Pull request closed without merging**